src/GraphQL/Resolvers/ObservableFromAsyncEnumerable.cs

```diff
@@ -0,0 +1,146 @@
+using System.Runtime.CompilerServices;
+using System.Security.Claims;
+using GraphQL.Execution;
+using GraphQL.Instrumentation;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Resolvers;
+
+/// <inheritdoc cref="ObservableFromAsyncEnumerable{T}.Create(Func{IResolveFieldContext, IAsyncEnumerable{T}})"/>
+internal sealed class ObservableFromAsyncEnumerable<T> : IObservable<object?>, IDisposable
+{
+    private readonly CancellationTokenSource _cancellationTokenSource;
+    private IAsyncEnumerable<T>? _enumerable;
+
+    private ObservableFromAsyncEnumerable(CancellationTokenSource cts, IAsyncEnumerable<T> enumerable)
+    {
+        _cancellationTokenSource = cts;
+        _enumerable = enumerable;
+    }
+
+    /// <summary>
+    /// Returns a source stream resolver delegate (which returns an <see cref="IObservable{T}"/>) from a delegate
+    /// which returns <see cref="IAsyncEnumerable{T}"/>. Each execution will create a new
+    /// <see cref="ObservableFromAsyncEnumerable{T}"/> instance which can only be used once.
+    /// </summary>
+    public static Func<IResolveFieldContext, ValueTask<IObservable<object?>>> Create(Func<IResolveFieldContext, ValueTask<IAsyncEnumerable<T>>> func)
+    {
+        return async context =>
+        {
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(context.CancellationToken);
+
+            // The underlying cancellation token is signaled if the HTTP connection is aborted. However, the
```

Comment:
Strictly speaking we don't know transport details here, but it's OK to work in such terms.